### PR TITLE
Support W3C distributed tracing standard

### PR DESCRIPTION
**Pull request by lmolkova**
Implements [w3c distributed tracing standard](https://github.com/w3c/distributed-tracing).

This change enables opt-in W3C support via settings.
To enable it, add W3COperationCorrelaitonTelemetryInitilizer and set `RequestTrackingTelemetryModule.EnableW3CHeadersExtraction` and `DependencyTrackingTelemetryModule.EnableW3CHeadersInjection` to true.

This change ensures W3C ids are set on the telemetry (operationId, parentId and Id). 'legacy' ids are set in custom dimensions whenever they are different from the W3C ones.

This change requires UI queries change to support querying for custom dimensions for the cases when operation id or parent id do not match on the legacy-W3C boundary.

-------------

**Out of date** please ignore
This is an experimental feature that could be turned on by setting `APPLICATIONINSIGHTS_ENABLE_W3C_TRACING` environment variable to `true`.

This is a temporary solution for non-production scenarios to validate W3C tracing standard.
It will be done properly within .NET SDK and Asp.NET frameworks.

For now, ApplicationInsights with enabled W3C support is not backward-compatible: i.e. correlation with other services instrumented with ApplicaitonInsights (but without the W3C) does NOT work.

### CODE REVIEW

**Comment by lmolkova** [comment]
@SergeyKanzhelev ping

File: Src/Common/W3C/W3CActivityExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    using System;
+    using System.ComponentModel;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.Linq;
+
+    /// <summary>
+    /// Extends Activity to support W3C distributed tracing standard.
+    /// </summary>
+    [Obsolete("Not ready for public consumption.")]
```
Comment (SergeyKanzhelev) [inline Src/Common/W3C/W3CActivityExtensions.cs]: will this Obsolete attribute apply to extension methods inside? I just don't know how it works for extensions...

File: Src/Common/W3C/W3CActivityExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    [EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (SergeyKanzhelev) [inline Src/Common/W3C/W3CActivityExtensions.cs]: Same question here - should it be applied to individual methods?

File: Src/Common/W3C/W3CActivityExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+        private static void SetVersion(this Activity activity, string value) =>
+            activity.AddTag(W3CConstants.VersionTag, value);
+
+        private static void SetSampled(this Activity activity, string value) =>
+            activity.AddTag(W3CConstants.SampledTag, value);
+
+        private static string GenerateSpanId()
+        {
+            // inefficient
```
Comment (SergeyKanzhelev) [inline Src/Common/W3C/W3CActivityExtensions.cs]: don't tell =)

File: Src/Common/W3C/W3CConstants.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        /// <summary>
+        /// Default version value.
+        /// </summary>
+        internal const string DefaultVersion = "00";
+
+        /// <summary>
+        /// Default sampled flag value.
+        /// </summary>
+        internal const string DefaultSampled = "01";
```
Comment (SergeyKanzhelev) [inline Src/Common/W3C/W3CConstants.cs]: I'm not sure about this one... We discussed `0` to be "decide for yourself" so maybe `0` is a better default.

File: Src/Common/W3C/W3CActivityExtensions.cs (hunk 2 above)
Comment (lmolkova) [reply Src/Common/W3C/W3CActivityExtensions.cs]: fixed

**Comment by lmolkova** [comment]
@SergeyKanzhelev @reyang @MS-TimothyMothra this is finally ready for the review. Please have a look

File: Src/Common/W3C/W3CActivityExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        [Obsolete("Not ready for public consumption.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public static void SetTraceParent(this Activity activity, string value)
+        {
+            if (value != null)
+            {
+                var parts = value.Split('-');
+                if (parts.Length == 4)
+                {
+                    activity.SetVersion(parts[0]);
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/W3C/W3CActivityExtensions.cs]: why is the order important?

File: Src/Common/W3C/W3CConstants.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+        /// <summary>
+        /// Sampled tag name.
+        /// </summary>
+        internal const string SampledTag = "w3c_sampled";
+
+        /// <summary>
+        /// TraceState tag name.
+        /// </summary>
+        internal const string TraceStateTag = "w3c_traceState";
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/W3C/W3CConstants.cs]: 1. this is internal stuff 
2. c# coding style suggest having const variables in PascalCase and we should follow specific language coding practices.

File: Src/Common/W3C/W3CActivityExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        [Obsolete("Not ready for public consumption.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public static string GetTraceParent(this Activity activity)
+        {
+            string version = null, traceId = null, spanId = null, sampled = null;
+            foreach (var tag in activity.Tags)
+            {
+                switch (tag.Key)
+                {
+                    case W3CConstants.TraceIdTag:
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/W3C/W3CActivityExtensions.cs]: why?

[the remark this answers is not in the document]
Comment (reyang) [reply Src/Common/W3C/W3CActivityExtensions.cs]: This would make it easier for people to read and maintain the code, similar like how we organize the using namespace statements - not a must have but definitely helpful.

File: Src/Common/W3C/W3CConstants.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (reyang) [reply Src/Common/W3C/W3CConstants.cs]: 1. I suggest that we treat both internal and external names the same way, for consistency.
2. Yes, we should follow PascalCase, and this is not the point here. The point is that in W3C we treat traceparent and tracestate as a "monolithic word", so the corresponding Pascal case should be Traceparent and Tracestate.

File: Src/Common/WebHeaderCollectionExtensions.cs [hunk 8]
```diff
@@ -55,8 +55,8 @@ public static void SetNameValueHeaderValue(this NameValueCollection headers, str
             Debug.Assert(headerName != null, "headerName must not be null");
             Debug.Assert(keyName != null, "keyName must not be null");
 
-            IEnumerable<string> headerValue = GetHeaderValue(headers, headerName);
-            headers[headerName] = string.Join(", ", HeadersUtilities.UpdateHeaderWithKeyValue(headerValue, keyName, value));
+            IEnumerable<string> headerValue = headers.GetHeaderValue(headerName);
+            headers[headerName] = string.Join(",", HeadersUtilities.UpdateHeaderWithKeyValue(headerValue, keyName, value));
```
[the remark this answers is not in the document]
Comment (reyang) [reply Src/Common/WebHeaderCollectionExtensions.cs]: let me rephrase my question, why do we do it in this pull request? Is this related to the W3C implementation, or it's some optimization we want to make?

[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/WebHeaderCollectionExtensions.cs]: we do this in this pull request because it's an obvious, nice and easy thing to do and improves things around tracestate implementation. I agree we may do it in the other PR, but being so subtle. it does not worth the hassle from my point of view

File: Src/Common/W3C/W3COperationCorrelationTelemetryInitializer.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                }
+
+                if (opTelemetry.Id != activity.Id)
+                {
+                    opTelemetry.Properties[W3CConstants.LegacyRequestIdProperty] = activity.Id;
+                }
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/W3C/W3COperationCorrelationTelemetryInitializer.cs]: we have a stylecop that checks such kind of issues and it did not complain. If you think having empty line at the end of file is useful, please modify the rule and apply it to all files :)

File: Src/Common/W3C/W3CActivityExtensions.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/W3C/W3CActivityExtensions.cs]: while this makes sense, I do have them ordered based on the importance and think this is an appropriate order.

File: Src/Common/StringUtilities.cs [hunk 10]
```diff
@@ -7,7 +7,12 @@
     /// <summary>
     /// Generic functions to perform common operations on a string.
     /// </summary>
-    public static class StringUtilities
+#if DEPENDENCY_COLLECTOR
+    public
```
Comment (SergeyKanzhelev) [inline Src/Common/StringUtilities.cs]: I like hiding this class from public surface. What's the reason to keep it public?

Comment (lmolkova) [reply Src/Common/StringUtilities.cs]: this class is used by AspNetCore SDK and should be visible to it. There are two alternatives:
- copy paste (as it is done now and i'm getting read of it by this PR)
- move common code to based SDK. I'm not fond of it because most of W3C code will go away from AI SDK eventually and I'd prefer to keep it all in one place. We can move other StringUtilities common methods in different PR

File: Src/Common/W3C/W3CConstants.cs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+        /// <summary>
+        /// W3C tracestate header name.
+        /// </summary>
+        public const string TraceStateHeader = "tracestate";
+
+        /// <summary>
+        /// Name of the field that carry ApplicationInsights application Id in the tracestate header.
+        /// </summary>
+        public const string ApplicationIdTraceStateField = "@msappid";
```
Comment (SergeyKanzhelev) [inline Src/Common/W3C/W3CConstants.cs]: recommendation is to add `@` for the vendor name in case of multi-tenant. For a single tenant - this delimiter is not needed.

Comment (SergeyKanzhelev) [reply Src/Common/W3C/W3CConstants.cs]: Also for azure we might need more than just app-id

Comment (lmolkova) [reply Src/Common/W3C/W3CConstants.cs]: ok, so do we need to add @ now? do we know what else we need right now?

File: Src/Common/W3C/W3CActivityExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+            {
+                var parts = value.Trim(' ', '-').Split('-');
+                if (parts.Length == 4 && !activity.IsW3CActivity())
+                {
+                    string traceId = parts[1];
+                    string parentSpanId = parts[2];
+
+                    byte.TryParse(parts[3], NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var sampled);
+
+                    if (traceId.Length == 32 && parentSpanId.Length == 16)
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply Src/Common/W3C/W3CActivityExtensions.cs]: the spec now says that app MAY ignore the traceid/spanid if they are not valid. I remember we discussed that implementation will have to validate traceis/spanid and ignore invalid ones.  I'll be happy to do this after spec will tell so :)
